storage/innobase/log/log0recv.cc

```diff
@@ -1612,6 +1612,22 @@ static byte *recv_parse_or_apply_log_rec_body(
   switch (type) {
 #ifndef UNIV_HOTBACKUP
     case MLOG_FILE_DELETE:
+      /* error out backup if undo truncation happens during backup */
+      if (srv_backup_mode && fsp_is_undo_tablespace(space_id) &&
+          backup_redo_log_flushed_lsn < recv_sys->recovered_lsn) {
+        ib::info() << "Last flushed lsn: " << backup_redo_log_flushed_lsn
+                   << " undo_delete lsn " << recv_sys->recovered_lsn;
+
+        ib::error(ER_IB_MSG_716)
+            << "An undo ddl truncation (could be automatic) "
+            << " operation has been "
+            << " performed. \n"
+            << " PXB will not be able to"
+            << " take a consistent backup."
+            << " Retry the backup"
+            << " operation later or with --lock-ddl";
+        exit(EXIT_FAILURE);
```

Comment:
@rahulmalik87 what happens if we have MLOG_FILE_DELETE on redo log before we start the backup and checkpoint mark is behind that? when we do the first redo log parse to catch-up the backup will abort even though it shouldn't?

Similar issue happens with MLOG_INDEX_LOAD, we first parse the redo logs via https://github.com/percona/percona-xtrabackup/pull/934/files#diff-29fd8beae6c2cfce456b51a213d966eeR4921 then we chage a boolean flag to indicates that from that point forward, parse an MLOG_INDEX_LOAD is unsafe, then we start the redo follow thread

Reply:
if condition's backup_redo_log_flushed_lsn < recv_sys->recovered_lsn
 checks ensures that we abort on MLOG_FILE_DELETE only for the redo generated after backup starts

Reply:
@rahulmalik87 I don't think that is true:
`backup_redo_log_flushed_lsn` is captured as part of parse_show_engine_innodb_status that happens on:
```
#0  parse_show_engine_innodb_status (connection=0x0) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/backup_mysql.cc:341
#1  0x00005555584a881e in xb_init () at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/xtrabackup.cc:6853
#2  0x00005555584aa661 in main (argc=6, argv=0x7fffffffe208) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/xtrabackup.cc:7455
```
Catch up of redo logs to parse everything that happened before the backup start until now happens at
```
#0  Redo_Log_Data_Manager::start (this=0x7fffffffd660) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/redo_log.cc:928
#1  0x000055555849f472 in xtrabackup_backup_func () at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/xtrabackup.cc:3894
#2  0x00005555584aa87c in main (argc=6, argv=0x7fffffffe208) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/xtrabackup.cc:7517
```
So first at main():7455 we get the point to start to parse the redo logs then later at main():7517 we do the catchup. The way it's designed now if we do a truncation before the backup start, first run of redo log will unnecessarily abort the backup.

Reply:
I think the catchup of redo logs before backup started is for a different reason. And this catchup will also include redo logs that are generated during the first scan of redo logs.

So noting the current flushed LSN at start of backup is really the point where redo logs are created at the time backup starts.

I do not get whats the problem here?

Reply:
@rahulmalik87 can you do a test:
Truncate an undo tablespace
start a backup

I think this will fail because truncation LSN is ahead of the checkpoint

Reply:
set global innodb_page_cleaner_disabled_debug=on;
CREATE UNDO TABLESPACE undo1 ADD DATAFILE 'undo1.ibu';
ALTER UNDO TABLESPACE innodb_undo_002 SET INACTIVE;

backup would be successful.

Reply:
```
(gdb) bt
+bt
#0  recv_parse_or_apply_log_rec_body (type=MLOG_FILE_DELETE, ptr=0x7ffff0969691 "", end_ptr=0x7ffff0a8076e "", space_id=4294967276, page_no=0, block=0x0, mtr=0x0, parsed_bytes=4) at /work/pxb/src/8.0/storage/innobase/log/log0recv.cc:1616
#1  0x00005555591d4b4f in recv_parse_log_rec (type=0x7fffffffcf7c, ptr=0x7ffff096968d "\243\373", <incomplete sequence \354>, end_ptr=0x7ffff0a8076e "", space_id=0x7fffffffcf84, page_no=0x7fffffffcf80, body=0x7fffffffcf88) at /work/pxb/src/8.0/storage/innobase/log/log0recv.cc:2746
#2  0x00005555591d4d2a in recv_single_rec (ptr=0x7ffff096968d "\243\373", <incomplete sequence \354>, end_ptr=0x7ffff0a8076e "") at /work/pxb/src/8.0/storage/innobase/log/log0recv.cc:2830
#3  0x00005555591d5571 in recv_parse_log_recs (checkpoint_lsn=20058523) at /work/pxb/src/8.0/storage/innobase/log/log0recv.cc:3110
#4  0x00005555585209ab in Redo_Log_Parser::parse_log (this=0x7fffffffd2a8, buf=0x7ffff6820000 "\200", len=2365440, start_lsn=20058112, checkpoint_lsn=20058523) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/redo_log.cc:418
#5  0x0000555558522f8e in Redo_Log_Data_Manager::copy_once (this=0x7fffffffd1e0, is_last=false, finished=0x7fffffffd0f7) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/redo_log.cc:1042
#6  0x0000555558522ace in Redo_Log_Data_Manager::start (this=0x7fffffffd1e0) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/redo_log.cc:949
#7  0x000055555849f472 in xtrabackup_backup_func () at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/xtrabackup.cc:3894
#8  0x00005555584aa87c in main (argc=6, argv=0x7fffffffdcf8) at /work/pxb/src/8.0/storage/innobase/xtrabackup/src/xtrabackup.cc:7517
```
backup parsed a MLOG_FILE_DELETE that happened before it started.

Reply:
Yes, that is because checkpoint didn't move.  We ddin't abort because we noted the last LSN that made to disk at start of backup.

So we tolerate all such MLOG_FILE_DELETS from checkpoint_LSN until the flushed_to_disk LSN noted (at start of backup)

Reply:
right, misunderstood it. LGTM